### ISSUE TITLE
Unable to read encfs config via named pipe

### ISSUE THREAD

**Issue by antofthy** [vgough/encfs#253]
For a VERY long time..  (Since 2009)  I have used a special system where I do not store the ".encfs5" with the encrypted data, but store it in a special (secure) database along with the keys to the encrypted data.
To pass the configuration file to encfs I create a named pipe which I put into the "ENCFS6_CONFIG" environment variable.   I then start the "encfs", which reads the config file from the named pipe as I write to it.   This way the config file is NEVER written to the disk in the clear.

For example -- (shell pseudo code - real code is written in perl)
   mkpipe temporary_named_pipe
   cat "the real config file info" >>  temporary_named_pipe &
   ENCFS6_CONFIG="temporary_named_pipe"  encfs  --stdin encfs_dir decoded_dir  <<<"$password"
   rm temporary_named_pipe

The problem is suddenly I get errors from encfs when it reads its config from a named pipe...

2016-12-16 20:59:43,610 ERROR [default] Failed to load config file /tmp/ks_lij9g9
2016-12-16 20:59:43,610 ERROR [default] Found config file /tmp/ks_lij9g9, but failed to load - exiting

If I replace the named pipe with a plain text file with the exact same config, but no other changes, it all works perfectly fine.   If interested the code is available in  "ks"  http://www.ict.griffith.edu.au/anthony/software/#ks

Why did it change?

 Why can't encfs read from a named pipe?

What can I do about it?

**Comment by rfjakob** [comment]
Running under strace (`ENCFS6_CONFIG="$PWD/pipe" strace encfs $PWD/a $PWD/b`) reveals this:

```
open("/tmp/tmp.9pa6Ln8X3Y/pipe", O_RDONLY) = 3
fstat(3, {st_mode=S_IFIFO|0664, st_size=0, ...}) = 0
lseek(3, 0, SEEK_SET)                   = -1 ESPIPE (Illegal seek)
read(3, "<?xml version=\"1.0\" encoding=\"UT"..., 4096) = 1264
fstat(3, {st_mode=S_IFIFO|0664, st_size=0, ...}) = 0
lseek(3, 0, SEEK_END)                   = -1 ESPIPE (Illegal seek)
lseek(3, 0, SEEK_CUR)                   = -1 ESPIPE (Illegal seek)
lseek(3, 0, SEEK_SET)                   = -1 ESPIPE (Illegal seek)
close(3)
```

The XML library tries to seek in the file, which fails, because pipes do not support seeking. Here is the code: https://github.com/leethomason/tinyxml2/blob/3.0.0/tinyxml2.cpp#L1832

Note that the XML library was recently changed to tinyxml2 ( https://github.com/vgough/encfs/commit/e78d1659e39e516ec4f0598ae2d91ddb704e95f2 )

**Comment by antofthy** [comment]
So it is being caused by a API that encfs is now using.  Seeks should be completely unnecessary for a straight forward read of data.  The only reason I can see it using seeks is that it is reading the file size so it can allocate the read buffer before actually reading!   A stupid restriction by "tinyxml2".

I have notice that the start of is now significantly slower, whether do to the API, or because I now need to
use a temporary file to hold the EncFS configuration.   This is the config now needs actual disk IO, rather than a memory buffer, as well as less parallel processing.  That I can no longer write into named piped memory, while encfs is still initialising.

**Comment by antofthy** [comment]
Why did you switch from using  boost libraries to TinyXML2?

**Comment by vgough** [comment]
Boost libraries had resulted in too many support issues due to version incompatibilities, and I didn't want to bundle boost as part of encfs.  So I removed them.

I'd be find with a PR to read the config file into memory before parsing.

**Comment by antofthy** [comment]
PR?

Reading the config into memory outside TinyXML is the main suggestion I am
seeing from that group.  It is after all what they are essentially doing.

I gave a suggestion to use  getdelim() (which is closely related to
getline()).  On a text file with a NULL character delimiter (instead of
return) getdelim() will read in the whole file, allocating and
re-allocating memory as needed.  You can then check for errors, zero size
before passing to TinyXML.

**Comment by rfjakob** [comment]
PR = github pull request

**Comment by antofthy** [comment]
Looks as though, TinyXML have accepted my suggestion to use getdelim() to
read in the file, and avoid the fseek() limitations.

**Comment by benrubson** [comment]
@antofthy could you close this as #319 has been merged ?
Many thx 👍

**Comment by benrubson** [comment]
@antofthy (or @rfjakob) could you close this as #319 has been merged ?
Many thx 👍

**Comment by antofthy** [comment]
Okay... when will it be released?  (version 1.9.2?)

**Comment by antofthy** [comment]
Thank you for fixing.. I hope it will be released soon!
